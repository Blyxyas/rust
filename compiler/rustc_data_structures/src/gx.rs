use std::{
    collections::{HashMap, HashSet},
    hash::BuildHasherDefault,
};

// pub use rustc_hash::{GxHashMap, GxHashSet, GxHasher};

pub use gxhash::GxHasher;

pub type StdEntry<'a, K, V> = std::collections::hash_map::Entry<'a, K, V>;

pub type GxHashMap<K, V> = HashMap<K, V, BuildHasherDefault<GxHasher>>;
pub type GxHashSet<T> = HashSet<T, BuildHasherDefault<GxHasher>>;

pub type GxIndexMap<K, V> = indexmap::IndexMap<K, V, BuildHasherDefault<GxHasher>>;
pub type GxIndexSet<V> = indexmap::IndexSet<V, BuildHasherDefault<GxHasher>>;
pub type IndexEntry<'a, K, V> = indexmap::map::Entry<'a, K, V>;
pub type IndexOccupiedEntry<'a, K, V> = indexmap::map::OccupiedEntry<'a, K, V>;

#[macro_export]
macro_rules! define_id_collections {
    ($map_name:ident, $set_name:ident, $entry_name:ident, $key:ty) => {
        pub type $map_name<T> = $crate::unord::UnordMap<$key, T>;
        pub type $set_name = $crate::unord::UnordSet<$key>;
        pub type $entry_name<'a, T> = $crate::gx::StdEntry<'a, $key, T>;
    };
}

#[macro_export]
macro_rules! define_stable_id_collections {
    ($map_name:ident, $set_name:ident, $entry_name:ident, $key:ty) => {
        pub type $map_name<T> = $crate::gx::GxIndexMap<$key, T>;
        pub type $set_name = $crate::gx::GxIndexSet<$key>;
        pub type $entry_name<'a, T> = $crate::gx::IndexEntry<'a, $key, T>;
    };
}
