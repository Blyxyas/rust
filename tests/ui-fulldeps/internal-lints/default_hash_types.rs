//@ compile-flags: -Z unstable-options

#![feature(rustc_private)]
#![deny(rustc::default_hash_types)]

extern crate rustc_data_structures;

use rustc_data_structures::gx::{GxHashMap, GxHashSet};
use std::collections::{HashMap, HashSet};

mod foo {
    pub struct HashMap;
}

fn main() {
    let _map: HashMap<String, String> = HashMap::default();
    //~^ ERROR prefer `GxHashMap` over `HashMap`, it has better performance
    //~^^ ERROR prefer `GxHashMap` over `HashMap`, it has better performance
    let _set: HashSet<String> = HashSet::default();
    //~^ ERROR prefer `GxHashSet` over `HashSet`, it has better performance
    //~^^ ERROR prefer `GxHashSet` over `HashSet`, it has better performance

    // test that the lint doesn't also match the Fx variants themselves
    let _fx_map: GxHashMap<String, String> = GxHashMap::default();
    let _fx_set: GxHashSet<String> = GxHashSet::default();

    // test another struct of the same name
    let _ = foo::HashMap;
}
