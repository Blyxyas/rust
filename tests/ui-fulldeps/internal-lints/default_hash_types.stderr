error: prefer `GxHashMap` over `HashMap`, it has better performance
  --> $DIR/default_hash_types.rs:16:41
   |
LL |     let _map: HashMap<String, String> = HashMap::default();
   |                                         ^^^^^^^
   |
   = note: a `use rustc_data_structures::gx::GxHashMap` may be necessary
note: the lint level is defined here
  --> $DIR/default_hash_types.rs:4:9
   |
LL | #![deny(rustc::default_hash_types)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^

error: prefer `GxHashMap` over `HashMap`, it has better performance
  --> $DIR/default_hash_types.rs:16:15
   |
LL |     let _map: HashMap<String, String> = HashMap::default();
   |               ^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: a `use rustc_data_structures::gx::GxHashMap` may be necessary

error: prefer `GxHashSet` over `HashSet`, it has better performance
  --> $DIR/default_hash_types.rs:19:33
   |
LL |     let _set: HashSet<String> = HashSet::default();
   |                                 ^^^^^^^
   |
   = note: a `use rustc_data_structures::gx::GxHashSet` may be necessary

error: prefer `GxHashSet` over `HashSet`, it has better performance
  --> $DIR/default_hash_types.rs:19:15
   |
LL |     let _set: HashSet<String> = HashSet::default();
   |               ^^^^^^^^^^^^^^^
   |
   = note: a `use rustc_data_structures::gx::GxHashSet` may be necessary

error: aborting due to 4 previous errors

